exp/api/v1alpha4/azuremanagedmachinepool_webhook.go

```diff
@@ -0,0 +1,179 @@
+
+var _ webhook.Validator = &AzureManagedMachinePool{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *AzureManagedMachinePool) ValidateCreate() error {
+	azuremanagedmachinepoollog.Info("validate create", "name", r.Name)
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (r *AzureManagedMachinePool) ValidateUpdate(oldRaw runtime.Object) error {
+	old := oldRaw.(*AzureManagedMachinePool)
+	var allErrs field.ErrorList
+
+	if r.Spec.SKU != old.Spec.SKU {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("Spec", "SKU"),
+				r.Spec.SKU,
+				"field is immutable"))
+	}
+
+	if old.Spec.OSDiskSizeGB != nil {
+		// Prevent OSDiskSizeGB modification if it was already set to some value
+		if r.Spec.OSDiskSizeGB == nil {
+			// unsetting the field is not allowed
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("Spec", "OSDiskSizeGB"),
+					r.Spec.OSDiskSizeGB,
+					"field is immutable, unsetting is not allowed"))
+		} else if *r.Spec.OSDiskSizeGB != *old.Spec.OSDiskSizeGB {
+			// changing the field is not allowed
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("Spec", "OSDiskSizeGB"),
+					*r.Spec.OSDiskSizeGB,
+					"field is immutable"))
+		}
+	}
+
+	if r.Spec.Mode != SystemNodePool && old.Spec.Mode == SystemNodePool {
+		// validate for last system node pool
+		if err := r.validateLastSystemNodePool(); err != nil {
+			allErrs = append(allErrs, field.Invalid(
+				field.NewPath("Spec", "Mode"),
+				r.Spec.Mode,
+				"Last system node pool cannot be mutated to user node pool"))
+		}
+	}
+
+	if len(allErrs) != 0 {
+		return apierrors.NewInvalid(GroupVersion.WithKind("AzureManagedMachinePool").GroupKind(), r.Name, allErrs)
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (r *AzureManagedMachinePool) ValidateDelete() error {
+	azuremanagedmachinepoollog.Info("validate delete", "name", r.Name)
+
+	if r.Spec.Mode != SystemNodePool {
+		return nil
+	}
+
+	return r.validateLastSystemNodePool()
+}
+
+func (r *AzureManagedMachinePool) validateLastSystemNodePool() error {
+	ctx := context.Background()
+
+	// Fetch the Cluster.
+	clusterName, ok := r.Labels[clusterv1.ClusterLabelName]
+	if !ok {
+		return nil
+	}
+
+	ownerCluster := &clusterv1.Cluster{}
+	key := client.ObjectKey{
+		Namespace: r.Namespace,
+		Name:      clusterName,
+	}
+
+	if err := webhookClient.Get(ctx, key, ownerCluster); err != nil {
+		if !azure.ResourceNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if !ownerCluster.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	opt1 := client.InNamespace(r.Namespace)
+	opt2 := client.MatchingLabels(map[string]string{
+		clusterv1.ClusterLabelName: clusterName,
+		LabelAgentPoolMode:         SystemNodePool,
```

Comment:
This is clever, but I'm curious how it works for a cluster in transition? Might need to add the label inside the reconciler for ammp created before this change went in to ensure they get it applied.

Reply:
I am not very clear what the doubt is here. 

The label is emeded into the ammp in the mutating webhook so i guess that should suffice and not require us to add labels in the reconciler.

Reply:
AMMP created before this commit and not updated after will not hit the webhook. They will be reconciled. For the webhook to work correctly, they all need to have the label. So it likely should be added inside reconciler.

Reply:
I guess there were no system pools before this commit, so that’s not really possible 😅 it’s fine as is, you are right